spec/requests/omniauth_callbacks_controller_spec.rb

```diff
@@ -33,6 +33,39 @@
       expect(Users::OmniauthCallbacksController.find_authenticator("twitter"))
         .not_to eq(nil)
     end
+
+    context "with a plugin-contributed auth provider" do
+      before do
+        @provider = Plugin::AuthProvider.new
+
+        @provider.authenticator = Auth::OpenIdAuthenticator.new('ubuntu', 'https://login.ubuntu.com', trusted: true)
+        @provider.enabled_setting = "ubuntu_login_enabled"
+
+        Discourse.stubs(:auth_providers).returns [ @provider ]
```

Comment:
I'm not sure if we can test this without stubbing, but we always prefer to do that if possible. It would be worth spending a few minutes looking at how hard that would be for your test.

Reply:
I think it is quite difficult without stubbing, but I may have missed something. 

`Discourse.auth_providers` populates from `Discourse.plugins`. I think the only way to populate `Discourse.plugins` is to run `Discourse.activate_plugins!`, which only pulls in plugins from the filesystem at `/plugins`.

Even the spec for `/lib/plugins/instance.rb` doesn't seem to populate `Discourse.plugins` so I think it would need some architectural changes to make this easier to test without stubbing.